File: src/providers/supabase/rules/non-atomic-replace-pattern.ts
/**
 * supabase-non-atomic-replace-pattern (correctness)
 *
 * Client-side delete-all-then-insert-all child rows without checking either
 * step's error can silently wipe data when a mid-loop insert fails.
 */
import { fromTableName, isSupabaseMutationKind, destructuredNames } from '../utils.js';

interface MutationSite {
  node: any;
  table: string | undefined;
  kind: 'delete' | 'insert';
  checksError: boolean;
}

function isSupabaseTableMutation(node: any, kind: 'delete' | 'insert'): boolean {
  return isSupabaseMutationKind(node, kind);
}

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Supabase delete-then-insert replace patterns should check errors or use RPC',
      category: 'correctness',
      rationale:
        'Replacing child rows by deleting all rows for a user then re-inserting is a common Supabase pattern, but sequential client calls are not transactional. If delete succeeds and a later insert fails, existing rows are gone with no error shown. Wrap both steps in a Postgres RPC function or check error after each call.',
      docsUrl: 'https://supabase.com/docs/guides/database/functions',
      recommended: true,
    },
    messages: {
      nonAtomicReplace:
        'This function deletes then re-inserts rows without checking errors — a failed insert after a successful delete loses data silently.',
    },
    schema: [],
  },
  create(context: any) {
    const fnStack: any[] = [];
    const mutationsByFunction = new Map<any, MutationSite[]>();

    function currentFunction() {
      return fnStack[fnStack.length - 1];
    }

    function recordMutation(node: any, awaitExpr: any, pattern: any | undefined, kind: 'delete' | 'insert') {
      const fn = currentFunction();
      if (!fn) return;
      if (!isSupabaseTableMutation(awaitExpr.argument, kind)) return;
      const checksError = pattern ? destructuredNames(pattern).has('error') : false;
      const list = mutationsByFunction.get(fn) ?? [];
      list.push({ node, table: fromTableName(awaitExpr.argument), kind, checksError });
      mutationsByFunction.set(fn, list);
    }

    function enterFunction(node: any) {
      fnStack.push(node);
    }

    function exitFunction(node: any) {
      const sites = mutationsByFunction.get(node);
      if (sites) {
        const deletes = sites.filter((s) => s.kind === 'delete');
        const inserts = sites.filter((s) => s.kind === 'insert');
        if (deletes.length > 0 && inserts.length > 0) {
          const uncheckedDelete = deletes.some((d) => !d.checksError);
          const uncheckedInsert = inserts.some((i) => !i.checksError);
          const sameTable =
            uncheckedDelete &&
            uncheckedInsert &&
            deletes.some((d) => inserts.some((i) => d.table && i.table && d.table === i.table));
          if (sameTable) {
            context.report({ node: inserts[0]!.node, messageId: 'nonAtomicReplace' });
          }
        }
        mutationsByFunction.delete(node);
      }
      fnStack.pop();
    }

    return {
      FunctionDeclaration(node: any) {
        enterFunction(node);
      },
      'FunctionDeclaration:exit'(node: any) {
        exitFunction(node);
      },
      FunctionExpression(node: any) {
        enterFunction(node);
      },
      'FunctionExpression:exit'(node: any) {
        exitFunction(node);
      },
      ArrowFunctionExpression(node: any) {
        enterFunction(node);
      },
      'ArrowFunctionExpression:exit'(node: any) {
        exitFunction(node);
      },
      VariableDeclarator(node: any) {
        if (node.init?.type !== 'AwaitExpression') return;
        const arg = node.init.argument;
        if (isSupabaseTableMutation(arg, 'delete')) recordMutation(node, node.init, node.id, 'delete');
        if (isSupabaseTableMutation(arg, 'insert')) recordMutation(node, node.init, node.id, 'insert');
      },
      ExpressionStatement(node: any) {
        const expr = node.expression;
        if (expr?.type !== 'AwaitExpression') return;
        const arg = expr.argument;
        if (isSupabaseTableMutation(arg, 'delete')) recordMutation(node, expr, undefined, 'delete');
        if (isSupabaseTableMutation(arg, 'insert')) recordMutation(node, expr, undefined, 'insert');
      },
    };
  },
};

export const supabaseNonAtomicReplacePatternRule = rule;
export default rule;

File: src/providers/supabase/rules/python/non_atomic_replace_pattern.py
"""supabase-non-atomic-replace-pattern

Parity with the JS rule, adapted for exceptions: sequential delete-then-insert
calls on the same table are not transactional in either SDK. In Python the
data-loss risk is specifically a *swallowed* exception on either step (see
`unchecked_mutation_error.py`) — an unwrapped call raises visibly and stops
execution, which is loud but not silent; a `try/except: pass` around either
step (commonly the whole delete+insert sequence sharing one wrapper) is what
lets a failed insert after a successful delete disappear without a trace.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import (
    build_parent_map,
    call_report_loc,
    enclosing_function,
    enclosing_try_noop_except,
    from_table_name,
    is_supabase_mutation_kind,
)

RULE_KEY = "supabase-non-atomic-replace-pattern"


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []
    parents = build_parent_map(tree)
    # function node -> list of (node, table, kind, unchecked)
    mutations_by_function: dict[ast.AST, list[tuple]] = {}

    for node in ast.walk(tree):
        if not isinstance(node, ast.Call):
            continue
        func = node.func
        if not (isinstance(func, ast.Attribute) and func.attr == "execute"):
            continue
        mutation_call = func.value
        if not isinstance(mutation_call, ast.Call):
            continue

        kind = None
        if is_supabase_mutation_kind(mutation_call, "delete"):
            kind = "delete"
        elif is_supabase_mutation_kind(mutation_call, "insert"):
            kind = "insert"
        if kind is None:
            continue

        fn = enclosing_function(parents, node)
        if fn is None:
            continue

        unchecked = enclosing_try_noop_except(parents, node)
        table = from_table_name(mutation_call)
        mutations_by_function.setdefault(fn, []).append((node, table, kind, unchecked))

    for fn, sites in mutations_by_function.items():
        deletes = [s for s in sites if s[2] == "delete"]
        inserts = [s for s in sites if s[2] == "insert"]
        if not deletes or not inserts:
            continue

        unchecked_delete = any(s[3] for s in deletes)
        unchecked_insert = any(s[3] for s in inserts)
        if not (unchecked_delete and unchecked_insert):
            continue

        same_table = any(
            d[1] and i[1] and d[1] == i[1] for d in deletes for i in inserts
        )
        if not same_table:
            continue

        report_node = inserts[0][0]
        line, col = call_report_loc(report_node)
        out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
